Implement BIP 8 based Speedy Trial activation

achow101 opened the pull request:
Implements the block height (BIP 8) based Speedy Trial activation proposal that was discussed on the bitcoin-dev mailing list.

In order to do so, this PR first changes the versionbits (BIP 9) implementation to use heights rather than MTP. Then the minimum activation height parameter is introduced. This also adds the minimum activation height as a third parameter of `-vbparams`. Additionally, the threshold is made a parameter of each deployment so that it can be customized for each deployment. Lastly several unit and functional tests are added.

In order to make this easier to backport, the names of functions and classes are largely unchanged (except where necessary for the height based things). This means that some things are misnamed as they refer to BIP 9 even though this is really BIP 8.

This PR borrows 2 commits from #19573 to implement height based versionbits, one commit from #21377 for tests, and 1 commit formerly part of #21380 for the threshold parameterization.

luke-jr committed:
Migrate versionbits to use height instead of MTP

Co-authored-by: Anthony Towns <aj@erisian.com.au>

luke-jr committed:
Rename user facing mentions of BIP 9 to versionbits and/or BIP 8

Co-authored-by: Anthony Towns <aj@erisian.com.au>

ajtowns committed:
tests: test versionbits delayed activation

achow101 committed:
Add minimum activation height to BIP9Deployments

benthecarman commented on src/validation.h:
```diff
@@ -226,7 +226,6 @@ struct MempoolAcceptResult {
 MempoolAcceptResult AcceptToMemoryPool(CChainState& active_chainstate, CTxMemPool& pool, const CTransactionRef& tx,
                                        bool bypass_limits, bool test_accept=false) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
 
-
```
nit: unneeded

achow101 replied:
Removed

benthecarman commented on src/chainparamsbase.cpp:
```diff
@@ -23,7 +23,7 @@ void SetupChainParamsBaseOptions(ArgsManager& argsman)
     argsman.AddArg("-segwitheight=<n>", "Set the activation height of segwit. -1 to disable. (regtest-only)", ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::DEBUG_TEST);
     argsman.AddArg("-testnet", "Use the test chain. Equivalent to -chain=test.", ArgsManager::ALLOW_ANY, OptionsCategory::CHAINPARAMS);
-    argsman.AddArg("-vbparams=deployment:start:end", "Use given start/end times for specified version bits deployment (regtest-only)", ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::CHAINPARAMS);
+    argsman.AddArg("-vbparams=deployment:@startheight:@timeoutheight", "Use given start/timeout heights for specified version bits deployment (regtest-only)", ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::CHAINPARAMS);
```
I think it'd be worth adding the ability to set the `min_activation_height` for testing as well

ajtowns replied:
As it stands, `-vbparams` is only used to disable taproot activation to check soft-fork status, so this would only be useful if it came with an actual test.

ajtowns commented:
Maybe consider rebasing on #21380 ; it adds fuzz testing and removes the need for the `MinActivationHeight()` virtual functions, and should make the "per-deployment-threshold" part a bit simpler.

ajtowns committed:
chainparams: make versionbits threshold per-deployment

achow101 commented:
> Maybe consider rebasing on #21380 ; it adds fuzz testing and removes the need for the `MinActivationHeight()` virtual functions, and should make the "per-deployment-threshold" part a bit simpler.

Done. It did indeed make this simpler. The fuzzer also caught one minor issue.

JeremyRubin commented on src/versionbits.cpp:
```diff
@@ -30,8 +36,8 @@ ThresholdState AbstractThresholdConditionChecker::GetStateFor(const CBlockIndex*
-        if (pindexPrev->GetMedianTimePast() < nTimeStart) {
-            // Optimization: don't recompute down further, as we know every earlier block will be before the start time
+        if (pindexPrev->nHeight + 1 < height_start) {
+            // Optimization: don't recompute down further, as we know every earlier block will be before the start height
```
there's not really a great way to rewrite this, but can we comment why it is +1

JeremyRubin replied:
how it is doc'd below is great

achow101 replied:
Done

JeremyRubin commented on src/versionbits.cpp:
```diff
@@ -94,50 +103,54 @@ ThresholdState AbstractThresholdConditionChecker::GetStateFor(const CBlockIndex*
+    stats.threshold = m_dep.m_threshold;
 
-    if (pindex == nullptr)
+    if (pindexPrev == nullptr || (pindexPrev->nHeight + 1) < stats.period) {
```
ibid -- a temporary makes this clearer

achow101 replied:
Done

JeremyRubin reviewed (commented):
generally looks ok to me couple minor comments; can do a more detailed review later

achow101 commented on src/consensus/params.h:
```diff
@@ -20,24 +20,37 @@ enum DeploymentPos
-    /** Timeout/expiry MedianTime for the deployment attempt. */
-    int64_t nTimeout;
+    /** Start block height for version bits miner confirmation. Must be a retarget block, can be in the past. */
+    int64_t startheight;
```
The potential for uninitialized members here is concerning to me, so I am working on a followup to deal with those. That may be rolled into this PR.

achow101 committed:
Clarify and reduce nRuleChangeActivationThreshold

As thresholds are now parameterized, nRuleChangeActivationThreshold is
no longer the threshold used for activating new rule changes. Instead it
is now only used to warn if there is an unkonwn versionbits deployment.
To make this clear, rename to m_vbits_min_threshold and update the
comment describing it.

Additionally, because this is just a minimum used for a warning, reduce
the threshold to 75% so that future soft forks which may have thresholds
lower than 95% will still have warnings.

Sjors commented on test/functional/feature_taproot.py:
```diff
@@ -1201,7 +1201,7 @@ def set_test_params(self):
         self.setup_clean_chain = True
         # Node 0 has Taproot inactive, Node 1 active.
-        self.extra_args = [["-par=1", "-vbparams=taproot:1:1"], ["-par=1"]]
+        self.extra_args = [["-par=1", "-vbparams=taproot:@1:@1"], ["-par=1"]]
```
While you're at it, the comment about which node is active is the wrong way around.

achow101 replied:
No? Node 0 has taproot inactive and setting the vbparams in this way (startheight=1, timeoutheight=1) ensures that it cannot activate on regtest.

Sjors replied:
No, taproot is always active on regtest: https://github.com/bitcoin/bitcoin/blob/968bc1e07b4a8a359876f5b8035fb4c369d04bad/src/chainparams.cpp#L418

This had me massively confused too, but you can tell from `testmempoolaccept` if you manually craft a (broken) taproot transaction: https://github.com/BitMEXResearch/forkmonitor/blob/master/spec/models/chaintip_spec.rb#L377 (`mempool_broken_tap` is created by node A, which has no extra_args: https://github.com/BitMEXResearch/forkmonitor/blob/master/spec/models/chaintip_spec.rb#L22)

Sjors replied:
Ah crap, was reading the brackets wrong. Nvm.

Sjors reviewed (commented):
Concept ACK. Did some light review of 9e50401c8a593c569f586786d84554c027fd8e8d.

Would love to see a functional test that shows the speedy trial in action, by looking at `softforks` in `getblockchaininfo` (I might write one myself)

Sjors committed:
test: add min_activation_height to -vbparams

Sjors committed:
test: BIP 8 delayed activation functional test

achow101 commented on doc/bips.md:
```diff
@@ -1,6 +1,7 @@
 BIPs that are implemented by Bitcoin Core (up-to-date up to **v0.21.0**):
 
-* [`BIP 9`](https://github.com/bitcoin/bips/blob/master/bip-0009.mediawiki): The changes allowing multiple soft-forks to be deployed in parallel have been implemented since **v0.12.1**  ([PR #7575](https://github.com/bitcoin/bitcoin/pull/7575))
+* [`BIP 8`](https://github.com/bitcoin/bips/blob/master/bip-0008.mediawiki): The changes for parallel, rapid deployment based on block height miner activation have been implemented since **v0.21.1** ([PR #21392](https://github.com/bitcoin/bitcoin/pull/21392)). The  with UASF failover is not yet implemented.
```
Done

achow101 commented:
> Here you go, a BIP 8 delayed activation functional test for your cherry-picking: https://github.com/Sjors/bitcoin/commits/2021/03/bip8-speedy-trial-test

Cherry picked.

Sjors commented:
Linter complains `'time' imported but unused`; my bad, I needed that import during an intermediate struggle (when generating coins on node 1, node 2 refused to sync).

achow101 commented:
Fixed the linter

Sjors reviewed (commented):
The code changes for 7ff01fb91960b029f67fde68de6a9ea0178c8ab3 look good to me, but I'm a bit confused by the changes in and removal of tests in d4c40a5.

achow101 commented on src/test/versionbits_tests.cpp:
```diff
@@ -178,65 +178,40 @@ BOOST_FIXTURE_TEST_SUITE(versionbits_tests, TestingSetup)
 BOOST_AUTO_TEST_CASE(versionbits_test)
 {
     for (int i = 0; i < 64; i++) {
-        // DEFINED -> FAILED
```
This change is from #19573. I believe it is dropped because BIP 8 specifies that the timeoutheight falls on a retarget block greater than the startheight, so this scenario is supposed to be impossible to occur. With BIP 9, it was possible because there was no guarantee of the number of blocks between the start time and timeout time, but with height, the number of blocks is guaranteed.

achow101 commented on src/test/versionbits_tests.cpp:
```diff
@@ -284,100 +259,67 @@ BOOST_AUTO_TEST_CASE(versionbits_params)
-    const auto chainParams = CreateChainParams(*m_node.args, CBaseChainParams::MAIN);
+    const auto period = CreateChainParams(*m_node.args, CBaseChainParams::REGTEST)->GetConsensus().nMinerConfirmationWindow;
+    gArgs.ForceSetArg("-vbparams", strprintf("testdummy:@%s:@%s", period, period * 2));
+    const auto chainParams = CreateChainParams(*m_node.args, CBaseChainParams::REGTEST);
```
I believe this change was done so that custom activation parameters could be used. Specifically, since mainnet start and timeout heights are in the hundreds of thousands of blocks, the test would just be wasting time building up a chain to reach that those heights. With regtest, it can use a way lower height so the test runs in a timely fashion.

Sjors replied:
I guess the better question is: why did it use mainnet in the past?

achow101 replied:
I guess it was convenient and doesn't really matter when the time can be faked?

ajtowns replied:
I figured the theory was "if you're going to test anything, test what's going to happen on mainnet". Since there's no proof of work on mainnet, it might be reasonable to test with real heights -- it's just building a skiplist with half a million entries to get to the start height, and going from there?

ajtowns replied:
Oh, no -- the reason was that it is possible to test "TESTDUMMY" activation on mainnet with MTP because you can fake the block time to be prior to the the genesis block's timestamp; but you can't do that with height because you obviously can't have a block height prior to NEVER_ACTIVE ie -2.

chewy917 reviewed (approved):
31fe80ae-50c8-4165-b57c-f65159b83877

kristapsk commented:
There are some compiler warnings for me (GCC 9.3.0).
```
test/fuzz/versionbits.cpp: In function ‘void versionbits_fuzz_target(FuzzBufferType)’:
test/fuzz/versionbits.cpp:150:37: warning: ‘*((void*)& last_stats +16)’ may be used uninitialized in this function [-Wmaybe-uninitialized]
  150 |                     if (last_stats->possible) {
      |                         ~~~~~~~~~~~~^~~~~~~~
test/fuzz/versionbits.cpp:149:97: warning: ‘last_stats.VBitsStats::count’ may be used uninitialized in this function [-Wmaybe-uninitialized]
  149 |                     assert(stats.count == last_stats->count || stats.count == last_stats->count + 1);
      |                                                                                                 ^
test/fuzz/versionbits.cpp:127:35: warning: ‘last_stats.VBitsStats::elapsed’ may be used uninitialized in this function [-Wmaybe-uninitialized]
  127 |                 assert(last_stats && last_stats->elapsed == period - 1);
      |                                   ^~
```

achow101 commented:
> There are some compiler warnings for me (GCC 9.3.0).

I believe these are caused by the changes from #21380.

ajtowns commented:
Hmm, I'm not seeing those warnings, and I think all those statements are protected by an `if (last_stats)` which should ensure it's initialised.

achow101 commented:
I see this warnings with gcc 10.2.0

ajtowns commented:
Apparently I wasn't seeing those warnings because my fuzz things were being compiled with `-O0`... Seems to be a gcc bug not handling optional very well to me ([80635](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=80635) maybe?) but it's easy enough to not use an optional here, so I've added a patch to do that.

luke-jr commented on src/chainparams.cpp:
```diff
@@ -189,22 +193,26 @@ class CTestNetParams : public CChainParams {
         consensus.CSVHeight = 770112; // 00000000025e930139bac5c6c31a403776da130831ab85be56578f3fa75369bb
         consensus.SegwitHeight = 834624; // 00000000002b980fcd729daaa248fd9316a5200e9b367f4ff2c42453e84201ca
-        consensus.MinBIP9WarningHeight = 836640; // segwit activation height + miner confirmation window
+        consensus.MinVBitsWarningHeight = 836640; // segwit activation height + miner confirmation window
```
We had discussed delaying the member/variable renames until after the actual logic changes for better backport/merge/review, and I thought that was a good idea. Are you planning to take that out of the PR? Been waiting to review pending on that...

achow101 replied:
Oh yeah, forgot to do that. I've removed the rename commit that I added in this PR, but #21380 also has a rename commit and it's up to @ajtowns to remove that if desired.

achow101 commented on doc/release-notes-21392.md:
```diff
@@ -0,0 +1,7 @@
+RPC
+---
+
+* BIP 9 has been replaced with a partial implementation of BIP 8. This change is reflected in `getblockchaininfo` where references to BIP 9 have been replaced with references to BIP 8.
```
`getblockchaininfo` has a `bip9` field that is changed to `bip8`. This only effects signet and regtest though since those are the only ones that have taproot show up in `getblockchaininfo`.

But this point is moot as I have removed the rename commit.

luke-jr commented on src/chainparams.cpp:
```diff
@@ -492,22 +509,25 @@ void CRegTestParams::UpdateActivationParametersFromArgs(const ArgsManager& args)
+                UpdateVersionBitsParameters(Consensus::DeploymentPos(j), start_height, timeout_height, min_activation_height);
                 found = true;
-                LogPrintf("Setting version bits activation parameters for %s to start=%ld, timeout=%ld\n", vDeploymentParams[0], nStartTime, nTimeout);
+                LogPrintf("Setting version bits activation parameters for %s to start_height=%ld, timeout_height=%ld\n", vDeploymentParams[0], start_height, timeout_height);
```
Missing the new `min_activation_height`

achow101 replied:
Done

luke-jr commented on src/chainparams.cpp:
```diff
@@ -492,22 +509,25 @@ void CRegTestParams::UpdateActivationParametersFromArgs(const ArgsManager& args)
-        if (vDeploymentParams.size() != 3) {
-            throw std::runtime_error("Version bits parameters malformed, expecting deployment:start:end");
+        if (vDeploymentParams.size() < 3 || vDeploymentParams.size() > 4) {
+            throw std::runtime_error("Version bits parameters malformed, expecting deployment:@start_height:@timeout_height[:@min_activation_height]");
```
I'm not sure it's a good idea to use underscores here when everything else uses `startheight`

achow101 replied:
Done

luke-jr commented on src/consensus/params.h:
```diff
@@ -22,22 +22,31 @@ enum DeploymentPos
+    /**
+     * If lock in occurs, delay activation until at least this block height. Activations only occur on retargets.
+     */
+    int64_t m_min_activation_height;
```
`{0}` here can avoid needing to initialise it everywhere it isn't used (and reduce the risk of ever forgetting to do so)

achow101 replied:
Done

luke-jr commented on src/versionbits.cpp:
```diff
@@ -5,15 +5,21 @@
+    int nThreshold = m_dep.threshold;
+    int64_t height_start = m_dep.startheight;
+    int64_t height_timeout = m_dep.timeoutheight;
+    int64_t min_activation_height = m_dep.m_min_activation_height;
```
`height_active_min` might be more consistent with the other variables.

achow101 replied:
Done

luke-jr commented on src/test/versionbits_tests.cpp:
```diff
@@ -26,35 +26,65 @@ static const std::string StateName(ThresholdState state)
+    TestConditionChecker() : ThresholdConditionChecker(m_dep_storage, 10)
+    {
+        m_dep_storage.bit = 8;
+        m_dep_storage.startheight= 100;
```
nit: space

achow101 replied:
Done

luke-jr commented on src/chainparamsbase.cpp:
```diff
@@ -23,7 +23,7 @@ void SetupChainParamsBaseOptions(ArgsManager& argsman)
     argsman.AddArg("-segwitheight=<n>", "Set the activation height of segwit. -1 to disable. (regtest-only)", ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::DEBUG_TEST);
     argsman.AddArg("-testnet", "Use the test chain. Equivalent to -chain=test.", ArgsManager::ALLOW_ANY, OptionsCategory::CHAINPARAMS);
-    argsman.AddArg("-vbparams=deployment:start:end", "Use given start/end times for specified version bits deployment (regtest-only)", ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::CHAINPARAMS);
+    argsman.AddArg("-vbparams=deployment:@startheight:@timeoutheight", "Use given start/timeout heights for specified version bits deployment (regtest-only)", ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::CHAINPARAMS);
```
`-vbparams` has been extended to accept min activation height, but the doc here isn't updated with it.

achow101 replied:
Done

luke-jr commented on src/versionbits.cpp:
```diff
@@ -5,15 +5,21 @@
+    int64_t min_activation_height = m_dep.m_min_activation_height;
+
+    // Check if this deployment is never active.
+    if (height_start == Consensus::VBitsDeployment::NEVER_ACTIVE && height_timeout == Consensus::VBitsDeployment::NEVER_ACTIVE ) {
```
nit: extra space

achow101 replied:
Fixed

michaelfolkson commented on test/functional/feature_bip8.py:
```diff
@@ -0,0 +1,126 @@
+        # any new rules apply to transactions currently in the mempool, which
+        # might be mined in the next block.
+        #
+        # The next retarteting period starts at block 144, so nothing should
```
nit: s/retarteting/retargeting

achow101 replied:
Fixed

luke-jr commented on test/functional/feature_bip8.py:
```diff
@@ -0,0 +1,126 @@
+        # Node 1 has a regular activation window
+        # Node 2 uses speedy trial.
+        self.extra_args = [
+            [f'-vbparams=testdummy:@1:@1'],
+            [f'-vbparams=testdummy:@144:@{144 * 3}'],
+            [f'-vbparams=testdummy:@144:@{144 * 2}:@{144 * 4}'],
```
(and below) `f''`should be avoided to make backport to 0.21 nicer

luke-jr commented on test/functional/feature_bip8.py:
```diff
@@ -0,0 +1,126 @@
+        # Node 1 has a regular activation window
+        # Node 2 uses speedy trial.
+        self.extra_args = [
+            [f'-vbparams=testdummy:@1:@1'],
+            [f'-vbparams=testdummy:@144:@{144 * 3}'],
+            [f'-vbparams=testdummy:@144:@{144 * 2}:@{144 * 4}'],
```
(and below) f''should be avoided to make backport to 0.21 nicer

achow101 replied:
Done

amitiuttarwar commented:
Now that #21380 has been revamped to only add a fuzz test (no refactors), probably makes sense to update this PR to match?

Also, can you update the OP description to match the commits? OP refers to 5 commits but the branch has 14 😛

achow101 commented:
> Now that #21380 has been revamped to only add a fuzz test (no refactors), probably makes sense to update this PR to match?

I have kept the refactoring commits and dropped the fuzzer commits, as the refactors were the reason this PR was based on #21380.

> Also, can you update the OP description to match the commits? OP refers to 5 commits but the branch has 14 stuck_out_tongue

Updated the description.

Sjors commented:
It's probably a good idea to extract 524b41ffdb67ebaf1538a39b88344e7d4118b4b8 to 414640fbaf8cf00e8e37451764a77cdec3fdd730 into a seperate PR, and perhaps also incorporate @luke-jr's #21399. That covers all the refactoring and generic-making. This PR, the BIP 9 variant #21377, a BIP 8 (lot=false) and a BIP 9 activation can all build on that.

achow101 commented:
I've dropped the refactoring.

ajtowns commented on src/test/versionbits_tests.cpp:
```diff
@@ -157,65 +157,40 @@ BOOST_FIXTURE_TEST_SUITE(versionbits_tests, TestingSetup)
-                           .Mine(14333, TestTime(30003), 0).TestActive().TestStateSinceHeight(4000)
-                           .Mine(24000, TestTime(40000), 0).TestActive().TestStateSinceHeight(4000)
+                           .Mine(99, TestTime(10000) - 1, 0x101).TestDefined().TestStateSinceHeight(0) // One second more and it would be started
+                           .Mine(100, TestTime(10000), 0x101).TestStarted().TestStateSinceHeight(100) // So that's what happens the next period
```
Comment is wrong

achow101 replied:
Fixed.

ajtowns commented on src/chainparamsbase.cpp:
```diff
@@ -22,7 +22,7 @@ void SetupChainParamsBaseOptions(ArgsManager& argsman)
     argsman.AddArg("-segwitheight=<n>", "Set the activation height of segwit. -1 to disable. (regtest-only)", ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::DEBUG_TEST);
     argsman.AddArg("-testnet", "Use the test chain. Equivalent to -chain=test.", ArgsManager::ALLOW_ANY, OptionsCategory::CHAINPARAMS);
-    argsman.AddArg("-vbparams=deployment:@startheight:@timeoutheight", "Use given start/timeout heights for specified version bits deployment (regtest-only)", ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::CHAINPARAMS);
+    argsman.AddArg("-vbparams=deployment:@startheight:@timeoutheight@min_activation_height", "Use given start, timeout, and minimum activation heights for specified version bits deployment (regtest-only)", ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::CHAINPARAMS);
```
Missing `:[]` around min_activation_height

achow101 replied:
Done

ajtowns commented on test/functional/feature_bip8.py:
```diff
@@ -0,0 +1,126 @@
+            self.log.debug('Node #{}...'.format(i))
+            info = node.getblockchaininfo()
+            assert_equal(info['blocks'], 0)
+            assert_equal(info["softforks"]["testdummy"]["bip8"]["status"], "defined")
```
This loop structure is repeated a bunch; might be clearer as:

```python
    def test_height(self, height, *status, mine=None):
        if height > self.height:
            assert mine is not None
            self.log.info(f"Test status at height {height}...")
            self.nodes[mine].generate(height - self.height)
            self.sync_blocks()
        elif height < self.height:
            assert mine is None
            self.log.info(f"Roll back to {height}...")
            old_block = self.nodes[0].getblockhash(height + 1)
            for node in self.nodes:
                node.invalidateblock(old_block)
        else:
            assert height == 0 and self.height == 0
            self.log.info(f"Test status at genesis...")

        self.height = height

        for (i, node), st in zip(enumerate(self.nodes), status):
            self.log.debug('Node #{}...'.format(i))
            info = node.getblockchaininfo()
            assert_equal(info['blocks'], height)
            assert_equal(info["softforks"]["testdummy"]["bip8"]["status"], st)

    def run_test(self):
        self.test_height(0, "defined", "defined", "defined")

        # BIP 8 state transitions from "defined" to "started" or "failed" after
        # the last block of the retargeting period has been mined. This means
        # any new rules apply to transactions currently in the mempool, which
        # might be mined in the next block.
        #
        # The next retargeting period starts at block 144, so nothing should
        # happen at 142 and the state should change at 143.

        self.test_height(144-2, "defined", "defined", "defined", mine=0)
        self.test_height(144-1, "failed", "started", "started", mine=0)

        self.log.info("Test status when not signalling...")
        self.test_height(144*2-1, "failed", "started", "failed", mine=0)
        self.test_height(144*3-1, "failed", "failed", "failed", mine=0)


        self.log.info("Test status when signalling...")
        # The new branch has unique block hashes, because of the signalling and
        # because generate uses a deterministic address that depends on the node
        # index.

        self.test_height(144-1, "failed", "started", "started")
        self.test_height(144*2-1, "failed", "locked_in", "locked_in", mine=2)
        self.test_height(144*3-1, "failed", "active", "locked_in", mine=2)
        self.test_height(144*4-1, "failed", "active", "active", mine=2)
```

achow101 replied:
Done

ajtowns commented on test/functional/feature_bip8.py:
```diff
@@ -0,0 +1,126 @@
+        # Node 1 has a regular activation window
+        # Node 2 uses speedy trial.
+        self.extra_args = [
+            ['-vbparams=testdummy:@1:@1'],
```
Note `@1:@1` gets interpreted equivalently to `@144:@144` and transitions from DEFINED to FAILED for the block at height 144.

achow101 replied:
Since this is for a node that has testdummy inactive, I've just changed this to `@-2:@-2`.

ajtowns replied:
I don't think that works? "never active" means the fork won't show up in getblockchaininfo (to avoid testdummy appearing on mainnet).

achow101 replied:
I work around that by allowing a status of `None` to indicate that it shouldn't show up in `getblockchaininfo`. The purpose of this node is to generate non-signaling blocks anyways, so the status of testdummy for it doesn't matter that much.

ajtowns commented on src/versionbits.cpp:
```diff
@@ -49,20 +56,19 @@ ThresholdState AbstractThresholdConditionChecker::GetStateFor(const CBlockIndex*
             case ThresholdState::DEFINED: {
-                if (pindexPrev->GetMedianTimePast() >= nTimeTimeout) {
+                if (height >= height_timeout) {
                     stateNext = ThresholdState::FAILED;
```
This doesn't match the BIP 8 spec which specifies that DEFINED always transitions to STARTED and never skips directly to FAILED.

achow101 replied:
This condition was actually caught by the fuzzer (back when this PR included it).

The scenario is that the timeoutheight is less than one period after the startheight. Without this, such a deployment would transition to the STARTED state, and then potentially to LOCKED_IN. But all of that after the timeoutheight has been passed. That doesn't make any sense and doesn't fit within what we are expecting in any tests.

I think we need to revisit this case in the BIP.

achow101 replied:
After discussing this with @luke-jr, I've removed this transition. Since there should never be a deployment that has the timeoutheight less than one period after the startheight, I've also added a sanity check to the unit tests for this case.

ajtowns replied:
Yes, the "always reach STARTED" issue was why there's a separate NEVER_ACTIVE setting, I think.

BIP8 does specify "timeoutheight must be at least 4096 blocks (2 retarget intervals) after startheight", so you could treat timeoutheight being reached from DEFINED as an extension to the spec and define it any way you like, similar to the always/never active cases. (The old bip8 code in #19573 didn't do that in order to match the spec in the bip as closely as possible to catch spec/implementation mismatches)

As far as the fuzz tester goes, following the bip8 spec exactly should mean something like:

``c++
switch (state) {
case STARTED:
    assert(startheight <= height);
    if (exp_state == STARTED) {
        assert(height <= timeoutheight);
        assert(blocks_sig < threshold);
    } else {
        assert(exp_state == DEFINED);
    }
    break;
case FAILED:
    assert(exp_state == STARTED || exp_state == FAILED);
    assert(never_active_test || height >= timeoutheight);
} 
```   

or alternatively extending it so you can skip STARTED would probably not change as much but might warrant adding a check along the lines of:

```c++
static int next_boundary(int n, int m) { return (n + m - 1) - ((n - 1) % m); }
if (next_boundary(timeoutheight, period) >= next_boundary(startheight, period)) {
    assert(state == DEFINED || state == FAILED);
    assert(exp_state == DEFINED || exp_state == FAILED);
}
```

ajtowns commented on src/chainparams.cpp:
```diff
@@ -501,21 +501,21 @@ void CRegTestParams::UpdateActivationParametersFromArgs(const ArgsManager& args)
-        if (!ParseInt64(vDeploymentParams[2], &nTimeout)) {
-            throw std::runtime_error(strprintf("Invalid nTimeout (%s)", vDeploymentParams[2]));
+        if (vDeploymentParams[2].empty() || vDeploymentParams[2].front() != '@' || !ParseInt64(vDeploymentParams[2].substr(1), &timeoutheight)) {
+            throw std::runtime_error(strprintf("Invalid timeoutheight (%s)", vDeploymentParams[2]));
```
Could consider checking that `startheight` and `timeoutheight` (and `min_activation_height`) are either on retarget boundaries or match `-1, max int` or `-2, -2`. (Values not on retarget boundaries will be rounded up implicitly by the versionbits logic)

achow101 replied:
Perhaps for a followup.

achow101 replied:
I've decided to add checking for these parameters.

achow101 commented:
Rebased for hidden merge conflict with the fuzzer. As such, the versionbits fuzzer is updated for heights and min activation height.

achow101 commented on src/test/fuzz/versionbits.cpp:
```diff
@@ -276,47 +261,51 @@ FUZZ_TARGET_INIT(versionbits, initialize)
-        assert(current_block->GetMedianTimePast() < checker.m_begin);
-        assert(current_block->GetMedianTimePast() < checker.m_end);
+        assert(never_active_test || height < checker.m_begin);
+        assert(never_active_test || height < checker.m_end);
```
I don't quite follow as to why the range for both startheight and timeoutheight should be ` [0, period*(max_periods-2)]`. I had set the range currently used in order to avoid the issue discussed in https://github.com/bitcoin/bitcoin/pull/21392#discussion_r597258950.

ajtowns replied:
For bip9/mtp versionbits the cache lookup loop will say "oh, we haven't reach start time yet, must still be DEFINED" which is inconsistent with the "DEFINED" check afterwards that says "if we reached timeout and start time at the same time, we go to FAILED instead of STARTED". If the fuzzer tracked the mtp for the previous period, you could capture that logic too, but that seemed a bit messy.

Because bip8 versionbits always goes DEFINED to STARTED it doesn't have that inconsistency, so checking a broader range of parameters is easy.

achow101 replied:
Done

achow101 commented on src/test/fuzz/versionbits.cpp:
```diff
@@ -334,12 +323,7 @@ FUZZ_TARGET_INIT(versionbits, initialize)
-        } else {
-            assert(exp_state == ThresholdState::FAILED);
-        }
+        assert(state == ThresholdState::DEFINED);
```
This was surprising to me as well. Since this commit was pulled from #19573, I just left it as it came from there.

achow101 commented on src/test/fuzz/versionbits.cpp:
```diff
@@ -276,47 +261,51 @@ FUZZ_TARGET_INIT(versionbits, initialize)
+        }
         break;
     case ThresholdState::ACTIVE:
+        assert(height >= checker.m_min_activation);
```
Done

ajtowns commented on src/test/fuzz/versionbits.cpp:
```diff
@@ -276,47 +261,51 @@ FUZZ_TARGET_INIT(versionbits, initialize)
     case ThresholdState::FAILED:
-        assert(current_block->GetMedianTimePast() >= checker.m_end);
+        assert(height >= checker.m_end);
         assert(exp_state != ThresholdState::LOCKED_IN && exp_state != ThresholdState::ACTIVE);
```
I think you should be able to make this:

```c++
    case ThresholdState::FAILED:
        assert(height >= checker.m_begin);
        assert(height >= checker.m_end);
        assert(exp_state == ThresholdState::FAILED || exp_state == ThresholdState::STARTED);
```

achow101 replied:
Done

Sjors reviewed (approved):
ACK 80fc57990b4682bad350e1e7aff0ea4db73fec31

Maybe wait with merging until BIP 8 is updated with `min_activation_height` support (see https://github.com/bitcoin/bips/pull/1080, https://github.com/bitcoin/bips/pull/1081)

achow101 commented on src/versionbits.cpp:
```diff
@@ -9,11 +9,16 @@ ThresholdState AbstractThresholdConditionChecker::GetStateFor(const CBlockIndex*
+    int height_timeout = TimeoutHeight(params);
+
+    // Check if this deployment is never active.
+    if (height_start == Consensus::BIP9Deployment::NEVER_ACTIVE && height_timeout == Consensus::BIP9Deployment::NEVER_ACTIVE) {
```
I think it is fine to leave as is.

Sjors replied:
I could imagine some confusion if some sets `height_start` to `NEVER_ACTIVE` but then sets `height_timeout` to some real date. But hopefully this code gets enough review nobody would actually do that.

achow101 replied:
I've added an additional check to the sanity check unit test that the `NEVER_ACTIVE` and `ALWAYS_ACTIVE` parameters are set correctly.

achow101 commented on src/chainparamsbase.cpp:
```diff
@@ -22,7 +22,7 @@ void SetupChainParamsBaseOptions(ArgsManager& argsman)
     argsman.AddArg("-segwitheight=<n>", "Set the activation height of segwit. -1 to disable. (regtest-only)", ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::DEBUG_TEST);
     argsman.AddArg("-testnet", "Use the test chain. Equivalent to -chain=test.", ArgsManager::ALLOW_ANY, OptionsCategory::CHAINPARAMS);
-    argsman.AddArg("-vbparams=deployment:start:end", "Use given start/end times for specified version bits deployment (regtest-only)", ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::CHAINPARAMS);
+    argsman.AddArg("-vbparams=deployment:@startheight:@timeoutheight", "Use given start/timeout heights for specified version bits deployment (regtest-only)", ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::CHAINPARAMS);
```
Added documentation for the always active and never active cases.

luke-jr replied:
Do we really need that level of detail in debug-only options?

Sjors replied:
It's confused me more than zero times.

achow101 commented on test/functional/feature_taproot.py:
```diff
@@ -1201,7 +1201,7 @@ def set_test_params(self):
         self.setup_clean_chain = True
         # Node 0 has Taproot inactive, Node 1 active.
-        self.extra_args = [["-par=1", "-vbparams=taproot:1:1"], ["-par=1"]]
+        self.extra_args = [["-par=1", "-vbparams=taproot:@-2:@-2"], ["-par=1"]]
```
Done

luke-jr commented on src/chainparams.cpp:
```diff
@@ -466,14 +474,57 @@ class CRegTestParams : public CChainParams {
+        return false;
+    }
+    if (startheight == Consensus::BIP9Deployment::ALWAYS_ACTIVE && timeoutheight != Consensus::BIP9Deployment::NO_TIMEOUT) {
+        error = strprintf("When startheight is %d, timeoutheight must be %d", Consensus::BIP9Deployment::ALWAYS_ACTIVE, Consensus::BIP9Deployment::NO_TIMEOUT);
```
This isn't true, and specifying NO_TIMEOUT is kinda annoying...?

achow101 replied:
Removed it.

luke-jr reviewed (changes requested):
utACK 9a8334212bfa557b06d300f49962b3a772b63837

One last bug could be fixed, but it doesn't really matter

achow101 commented on src/chainparams.cpp:
```diff
@@ -466,14 +474,54 @@ class CRegTestParams : public CChainParams {
+        return false;
+    }
+    if (timeoutheight <= startheight) {
+        error = strprintf("Invalid timeoutheight (%d), must be greater than the startheight (%d)", timeoutheight, timeoutheight);
```
Fixed

luke-jr reviewed (approved):
utACK aa41d035137c0686ae3488e2f9e774c4edbf5bd7

sipa commented on src/chainparams.cpp:
```diff
@@ -474,14 +474,53 @@ class CRegTestParams : public CChainParams {
     void UpdateActivationParametersFromArgs(const ArgsManager& args);
 };
 
+bool CRegTestParams::CheckVBitsHeights(std::string& error, int64_t startheight, int64_t timeoutheight)
```
In commit "Migrate versionbits to use height instead of MTP", but more general perhaps.

There seems to be some overlap here with `sanity_check_params`. Would it make sense to have them merged, and perhaps make it always performed (also for mainnet/testnet, at startup)?

luke-jr replied:
Seems best left for a refactor later IMO.

achow101 replied:
I think it is reasonable to have the sanity check always performed at startup, but that can be left for a later PR.

However I have removed the overlap with `sanity_check_params` and have that instead call `CheckVBitsHeights` (now part of `CChainParams`) in the test for the height sanity checks.

achow101 commented on src/test/fuzz/versionbits.cpp:
```diff
@@ -135,38 +124,25 @@ FUZZ_TARGET_INIT(versionbits, initialize)
-        // this may make timeout too early; if so, don't run the test
-        if (start_time > timeout) return;
+        startheight = fuzzed_data_provider.ConsumeIntegralInRange<int>(0, period * (max_periods - 2));
+        timeoutheight = fuzzed_data_provider.ConsumeIntegralInRange<int>(0, period * (max_periods - 2));
```
afaict, the fuzzer is flexible enough to deal with that case. See also: https://github.com/bitcoin/bitcoin/pull/21392#discussion_r598061975

luke-jr commented on src/chainparams.cpp:
```diff
@@ -474,14 +474,53 @@ class CRegTestParams : public CChainParams {
+        error = strprintf("When one of startheight or timeoutheight is %d, both must be %d", Consensus::BIP9Deployment::NEVER_ACTIVE, Consensus::BIP9Deployment::NEVER_ACTIVE);
+        return false;
+    }
+    if (startheight == Consensus::BIP9Deployment::ALWAYS_ACTIVE) return true;
```
That was originally required, but unnecessary and since NO_TIMEOUT is a maxint, seems annoying to require for no reason.

achow101 replied:
Oops. The requirement was intentionally removed, but I didn't notice I had that earlier check. I've removed this line and dropped `NO_TIMEOUT` from the earlier check.

benthecarman reviewed (commented):
ACK 2cb41863abe59717a199a66030185e2d23226a23

Just a couple nits

achow101 commented on src/consensus/params.h:
```diff
@@ -25,19 +25,28 @@ enum DeploymentPos
+    /**
+     * If lock in occurs, delay activation until at least this block height. Activations only occur on retargets.
+     */
+    int64_t m_min_activation_height{0};
```
It should be `int` here, fixed.

For vbparams, the string to int parser returns an int64_t, so that's what I use there.

benthecarman commented on src/versionbits.cpp:
```diff
@@ -49,20 +57,17 @@ ThresholdState AbstractThresholdConditionChecker::GetStateFor(const CBlockIndex*
-                    break;
-                }
                 // We need to count
                 const CBlockIndex* pindexCount = pindexPrev;
```
By starting here, are we then ignoring if the latest block is signaling?

context:
```
// We track state by previous-block, so the height we should be comparing is +1
const int64_t height = pindexPrev->nHeight + 1;
```

benthecarman replied:
Also for my sanity, is the reason the height check is moved from the beginning to the end because we are using `pindexPrev->nHeight + 1`

achow101 replied:
Yes. Determining whether the threshold is reached is done at the beginning of each signaling period, and determines the state for the entire signaling period. This is checking for whether the threshold is reached in the previous signaling period to determine whether the current signaling period (starting at block `height`) is locked_in. The current block (at `height`) is not part of the previous signaling period, so it is not counted.

ajtowns replied:
Consider a block B just after a retarget period boundary, ie `B->nHeight % 2016 == 0`, and `pindexPrev = B->pprev`. `pindexPrev` was in state `STARTED`. What state should `B` be in? It depends on the signalling of the prior 2016 blocks -- so we test `pindexPrev->nVersion`, `pindexPrev->pprev->nVersion`, etc. If enough signal, we switch to `LOCKED_IN`. If they don't, we check `B->nHeight` (which is `B->prev->nHeight + 1` aka `pindexPrev->nHeight + 1`) to see if it's reached the timeout, and if so switch to `FAILED`.

The reason the height check is moved to the end is so that every retarget period that begins in `STARTED` is allowed to transition to `LOCKED_IN` -- previously, if the final period passed the timeout MTP midway through, it would be too late and would unconditionally transition to `FAILED`.

achow101 replied:
> Also for my sanity, is the reason the height check is moved from the beginning to the end of this case because we are using `pindexPrev->nHeight + 1`

No. The height check is moved because BIP 8 specifies that the threshold being reached takes precedence over the timeoutheight being reached. This is because determining whether the threshold is reached would be done at `timeoutheight`. So if the height check came first, the last signaling period would always fail. Instead, we want to succeed if the threshold is reached in the  last signaling period.

benthecarman reviewed (approved):
reACK 4a37423700f3432b976ada6adbc19e1ce75c7d9a

achow101 commented:
Had to rebase due to fuzzer changes

benthecarman commented:
reACK 2f457ca73ce33557ecff0d45c37c33957b1b0764

MarcoFalke commented on src/chainparams.cpp:
```diff
@@ -474,14 +474,51 @@ class CRegTestParams : public CChainParams {
+{
+    // Special always or never active cases
+    if ((startheight == Consensus::BIP9Deployment::NEVER_ACTIVE && timeoutheight == Consensus::BIP9Deployment::NEVER_ACTIVE)
+        || startheight == Consensus::BIP9Deployment::ALWAYS_ACTIVE) {
```
Any reason to not enforce NO_TIMEOUT here?

luke-jr replied:
Not well-defined, not very friendly (it's  a max int value), and ignored anyway. (third time addressing this)

achow101 replied:
In `GetStateFor`, we are checking that both `startheight` and `timeoutheight` are `NEVER_ACTIVE` in order to return the correct state for never active. It only checks `startheight` is `ALWAYS_ACTIVE` when determining always active. This behavior came from #19573 and I didn't change it. Perhaps @luke-jr or @ajtowns could explain why that is.

luke-jr replied:
It's simply a non-change from the original BIP9 code.

I don't see a reason to change it here.

MarcoFalke replied:
ALWAYS and NEVER are not specified in any BIP, they are special magic values in our code only (introduced in this pull request). Treating them in the same fashion will be less code, and less code complexity.

luke-jr replied:
NO_TIMEOUT is not introduced in this PR. Feel free to refactor in a followup... or achow can do it here if you care so strongly. I don't see that it matters either way.

achow101 replied:
I think we can leave this for a followup.

MarcoFalke commented on src/chainparams.cpp:
```diff
@@ -474,14 +474,51 @@ class CRegTestParams : public CChainParams {
-        consensus.vDeployments[d].nStartTime = nStartTime;
-        consensus.vDeployments[d].nTimeout = nTimeout;
+        consensus.vDeployments[d].startheight = startheight;
+        consensus.vDeployments[d].timeoutheight = timeoutheight;
```
I know this is just regtest, but an integral type of size 64-bit will be truncated if assigned to an integral type of size 32-bit

luke-jr replied:
Maybe have CheckVBitsHeights check this

MarcoFalke commented on src/test/fuzz/versionbits.cpp:
```diff
@@ -25,13 +25,13 @@ class TestConditionChecker : public AbstractThresholdConditionChecker
+    const int m_end = 0;
+    const int m_period = 0;
+    const int m_threshold = 0;
+    const int m_bit = 0;
```
Why is this changed?

achow101 replied:
Bad rebase conflict resolution, reverted.

maflcko reviewed (commented):
Looked at the first commit, I'll review the remaining commits later

devrandom commented:
Tested as follows:

- made unit tests fail by manually breaking the unit test `versionbits_test` in various ways
- made unit tests fail by introducing off-by-one errors in all the relevant spots in `versionbits.cpp`
- ensure no failure if cache is disabled by removing cache store statements in `versionbits.cpp`
- manually tested with a single local regtest node
- manually tested a reorg with two local regtest nodes going from `STARTED` to `LOCKED_IN` and back to `STARTED` when the reorg removes a single signalling block that brought the deployment just over the threshold

ariard commented on src/versionbits.cpp:
```diff
@@ -74,12 +79,16 @@ ThresholdState AbstractThresholdConditionChecker::GetStateFor(const CBlockIndex*
                 if (count >= nThreshold) {
                     stateNext = ThresholdState::LOCKED_IN;
+                } else if (height >= height_timeout) {
+                    stateNext = ThresholdState::FAILED;
```
Do you have test coverage for threshold taking precedence over reaching timeout ? This diff doesn't break `feature_bip8.py`

```
@@ -77,10 +78,11 @@ ThresholdState AbstractThresholdConditionChecker::GetStateFor(const CBlockIndex*
                     }
                     pindexCount = pindexCount->pprev;
                 }
+                if (height >= height_timeout) {
+                    stateNext = ThresholdState::FAILED;
+                }
                 if (count >= nThreshold) {
                     stateNext = ThresholdState::LOCKED_IN;
-                } else if (height >= height_timeout) {
-                    stateNext = ThresholdState::FAILED;
                 }
                 break;
             }
```

achow101 replied:
A unit test fails, although not because it explicitly tests this behavior. I've added such a test to feature_bip8.py.

However this particular diff wouldn't fail because it still moves to `LOCKED_IN` when the threshold is reached. You need to use `else if` in order for the `height >= height_timeout` check to take precedence.

ariard commented on src/versionbits.h:
```diff
@@ -24,10 +24,10 @@ static const int32_t VERSIONBITS_NUM_BITS = 29;
-    STARTED,   // For blocks past the starttime.
+    STARTED,   // For blocks past the startheight.
     LOCKED_IN, // For one retarget period after the first retarget period with STARTED blocks of which at least threshold have the associated bit set in nVersion.
     ACTIVE,    // For all blocks after the LOCKED_IN retarget period (final state)
```
I think you should document "For all blocks after the LOCKED_IN retarget period, if min activation height has been reached" ?

achow101 replied:
Done

ariard commented on src/rpc/blockchain.cpp:
```diff
@@ -1296,24 +1295,24 @@ RPCHelpMan getblockchaininfo()
-                                    {RPCResult::Type::NUM_TIME, "start_time", "the minimum median time past of a block at which the bit gains its meaning"},
-                                    {RPCResult::Type::NUM_TIME, "timeout", "the median time past of a block at which the deployment is considered failed if not yet locked in"},
+                                    {RPCResult::Type::NUM, "startheight", "the minimum height of a block at which the bit gains its meaning"},
+                                    {RPCResult::Type::NUM, "timeoutheight", "the height of a block at which the deployment is considered failed if not yet locked in"},
```
Maybe display "min_activation_height" if relevant (!= 0) ?

achow101 replied:
Changed to output it always.

sipa commented on src/test/fuzz/versionbits.cpp:
```diff
@@ -281,47 +260,46 @@ FUZZ_TARGET_INIT(versionbits, initialize)
     case ThresholdState::STARTED:
-        assert(current_block->GetMedianTimePast() >= checker.m_begin);
-        assert(current_block->GetMedianTimePast() < checker.m_end);
+        assert(height >= checker.m_begin);
```
In commit "Migrate versionbits to use height instead of MTP":

You could incorporate the later `if (never_active_test) assert(state == ThresholdState::DEFINED);` here as `assert(!never_active_test);`.

(I found this by drawing the finite state machine based on these tests, and noticed this line was missing before noticing it's done explicitly later)

achow101 replied:
Added the new assert, but I did not remove the later assert.

ariard commented on src/chainparams.cpp:
```diff
@@ -466,14 +474,60 @@ class CRegTestParams : public CChainParams {
+        return false;
+    }
+    if (timeoutheight < startheight + (int)consensus.nMinerConfirmationWindow) {
+        error = strprintf("Invalid timeoutheight (%d), must be at least one period greater than the startheight (%d)", timeoutheight, startheight);
```
According to BIP 8, "timeoutheight must be at least 4096 blocks (2 retarget intervals) after startheight" ?

achow101 replied:
Done.

sipa commented on src/test/fuzz/versionbits.cpp:
```diff
@@ -281,47 +260,46 @@ FUZZ_TARGET_INIT(versionbits, initialize)
-        assert(exp_state != ThresholdState::LOCKED_IN && exp_state != ThresholdState::ACTIVE);
+        assert(height >= checker.m_begin);
+        assert(height >= checker.m_end);
+        assert(exp_state == ThresholdState::FAILED || exp_state == ThresholdState::STARTED);
```
In commit "Migrate versionbits to use height instead of MTP":

I believe you should do `assert(exp_state == ThresholdState::FAILED || (exp_state == ThresholdState::STARTED && blocks_sig < threshold));` here instead. As stated, nothing prevents the FSM implementation from going from STARTED->FAILED when the threshold is reached in the last period.

achow101 replied:
Done

ariard commented on src/chainparams.cpp:
```diff
@@ -466,14 +474,60 @@ class CRegTestParams : public CChainParams {
+    if (timeoutheight < startheight + (int)consensus.nMinerConfirmationWindow) {
+        error = strprintf("Invalid timeoutheight (%d), must be at least one period greater than the startheight (%d)", timeoutheight, startheight);
+        return false;
+    }
```
I think BIP 8 implies that if min_activation_height is set, it must be superior to start_height. Though not explicit "minimum_activation_height should be set to several retarget periods in the future if the startheight is to be very soon after software with parameters is expected to be released."

achow101 replied:
The minimum activation height can be any value, it does not need to be greater than the start_height. If it is less than or equal to the start height, it just means that the locked_in period does not become extended.

ariard reviewed (commented):
Had a first parse on it

achow101 commented on src/chainparams.h:
```diff
@@ -116,6 +116,9 @@ class CChainParams
     const ChainTxData& TxData() const { return chainTxData; }
+
+    /** Check that the given heights make sense */
+    bool CheckVBitsHeights(std::string& error, int startheight, int timeoutheight, int min_actvation_height) const;
```
Fixed

achow101 commented on src/consensus/params.h:
```diff
@@ -25,19 +25,28 @@ enum DeploymentPos
+    int startheight;
+    /** Timeout/expiry block height for the deployment attempt. Must be a retarget block. */
+    int timeoutheight;
+    /** Threshold for activation */
```
Expanded this comment.

achow101 commented on src/chainparams.h:
```diff
@@ -116,6 +116,9 @@ class CChainParams
     const ChainTxData& TxData() const { return chainTxData; }
+
+    /** Check that the given heights make sense */
+    bool CheckVBitsHeights(std::string& error, int startheight, int timeoutheight) const;
```
Moving it to `versionbits.{h,cpp}` requires adding `versionbits.cpp` to `libbitcoin_common` and this introduces a bunch of other build system issues that I don't want to figure out. I've moved the function out of `CChainParams` though so that it could be moved elsewhere in the future as a move-only.

sipa commented:
Code review ACK for all but the unit and functional tests (which I've only casually looked over). I inferred the finite state machine diagram based on the fuzz test code (making a few suggestions), and compared that with the implemented logic.

Sjors commented:
re-ACK 2e55bcedb8d73e49620a5731196bf7e23bb53ccc

michaelfolkson commented:
ACK 2e55bcedb8d73e49620a5731196bf7e23bb53ccc

Ideally a choice on a Speedy Trial PR would have been made by now so I could focus review and testing on one PR and one approach. Spreading review over two competing PRs is doing none of us any favors. I would have liked to have spent more time testing this PR before ACKing it but I think the biggest danger at this point is spreading review too thinly across two PRs.

JeremyRubin commented on src/consensus/params.h:
```diff
@@ -25,19 +25,28 @@ enum DeploymentPos
+    int startheight;
+    /** Timeout/expiry block height for the deployment attempt. Must be a retarget block. */
+    int timeoutheight;
+    /** Threshold for lockin. Must be at least nRuleChangeActivationThreshold for that network. */
```
I think this variable name changed.

achow101 replied:
Will fix if I retouch.

luke-jr commented on src/rpc/blockchain.cpp:
```diff
@@ -1243,8 +1242,9 @@ static void BIP9SoftForkDescPushBack(UniValue& softforks, const std::string &nam
-    bip9.pushKV("timeout", consensusParams.vDeployments[id].nTimeout);
+    bip9.pushKV("startheight", consensusParams.vDeployments[id].startheight);
+    bip9.pushKV("timeoutheight", consensusParams.vDeployments[id].timeoutheight);
+    bip9.pushKV("minimum_activation_height", consensusParams.vDeployments[id].m_min_activation_height);
```
Probably should either hide this when 0, or use starttime+2*period

achow101 replied:
I think it's fine to always display.

achow101 commented on src/chainparams.cpp:
```diff
@@ -466,14 +474,60 @@ class CRegTestParams : public CChainParams {
+        return false;
+    }
+    if (timeoutheight < startheight + (2 * (int)consensus.nMinerConfirmationWindow)) {
+        error = strprintf("Invalid timeoutheight (%d), must be at least two periods greater than the startheight (%d)", timeoutheight, startheight);
```
Enforces this sentence in BIP 8:

> timeoutheight must be at least 4096 blocks (2 retarget intervals) after startheight.

ajtowns replied:
That's needed for lot=true/MUST_SIGNAL to work consistently with lot=false. Could also be solved by adding a transition directly from DEFINED to MUST_SIGNAL (skipping STARTED) I think.

luke-jr replied:
Nah, probably not worth changing. One period is probably crazy anyway. And if we ever do need it, we can add the extra logic for it at that time.

harding commented:
As of commit 2e55bcedb8d73e49620a5731196bf7e23bb53ccc, lightly tested the likely outcomes of ST and confirmed that `getblockchainfo` returned the expected values whether activation was achieved or not, before and after the minimum_activation_period.  I did not check close enough to detect off-by-one errors or other small details.

Tested creating a regtest chain with this branch and `-vbparams=taproot:@144:@576:@1008`, producing enough blocks to lockin and activate taproot, synced them to a taproot-oblivious 0.20.1 node, then upgraded that node to this branch with the same vbparams and confirmed that the upgraded node's `getblockchaininfo` results indicated it would be enforcing taproot.

I skimmed the code and didn't see any problems, but I don't feel confident enough in my reviewing abilities to say it definitely is fine.

maflcko reviewed (commented):
I looked at 2e55bcedb8d73e49620a5731196bf7e23bb53ccc 🚴

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

I looked at 2e55bcedb8d73e49620a5731196bf7e23bb53ccc 🚴
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUg82Av/XqMgcTbyFlMastLZ9AJdjHOHQjwpOkce6XC8ktfvqwbWpRioWiSHf0+C
7bN6kFEJMxXPUAOZ4DeHLoBrsCMqTwPJZ21nNV9p8ZhByBBp5KPEA2PJ3s/RwBCp
YeRpKlLj856117mUv/5oFnABj4xuUMmZ3OkwuvU3RN7TrfKBUkR8ZkmzoSaOkv7s
O/bnZC4qudRl/dGFoa+CouYMWpi49mT+kB59Rs8UZha86KH5Td2siTAbbJBiMAi5
8bFUwGSEsvxJGAr+52PHAcUUymGMZp29neY4Q5f7HIJ28uX5ged/EzSiFeeu4GZn
tH9B7mSqOvjnFIdOCvy8HYy8TzShKGvkiD4PcNvcVWfkQi2WbXtns1AjaZ5Tjeu9
ry2OHAzyWSvqbT1DIPD36VTb2c46u/vd9d05rkCjuccHcnXPkV49f68JblA2p9Fj
LUpoCswXIJPZtPxs2ZSwb0LxP54YW1RnguVOy6jg7PcMnecDeu/LFD4MsJtVJE9W
bqp2Q82X
=zjni
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `9fd4467b454c576b05f9c797bd20cd8620ad58c2b63ca8dd4e36f27ec7059a14  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e8929401089fd4467b454c576b05f9c797bd20cd8620ad58c2b63ca8dd4e36f27ec7059a14f0108ff042165174926429b160e1170de99108fff010181b185758ce789074becb639c35141a08f104605db9e8f008d5492509f582efee0083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff0102167bacd522665f9eca18f7a289ef8bb08f1209c69bb53b915473e27aef5c9b98e92d9e3016d4a97e48329e3d1cb3139462d3f08f104605db9e7f0081578f1c2ee261cf00083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff0102a752c38f9010d844d40ec87e0945efc08f120e8ba046c0af1a9d5891f26cea4a67e9fb0da321a1982bbf3398e9a8ac652a53708f1209424495065a00647bd33989bdc09615acbb5e1785a25c0e98a29154bf8b546c808f104605dac10f0083a35f7be1fc29ee80083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6df0105064ade08e1e6f34c85dab49996b070408f02008dd19a2586e27e080b270b3b32ed0e9a126e04602bf60b0005d16c37714a61d08f1204b7b124a5b9b779bb7419b145138cb957aa8ef16c554fbb7525711b7a8656c2e08f104605db9e6f008ff5c55a3b32230ea0083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

devrandom commented:
ACK https://github.com/bitcoin/bitcoin/pull/21392/commits/2e55bcedb8d73e49620a5731196bf7e23bb53ccc

In addition to the testing in https://github.com/bitcoin/bitcoin/pull/21392#issuecomment-805910099 I also checked that the fuzzer detects various off-by-one errors I manually introduced into the implementation.

luke-jr reviewed (approved):
utACK 2e55bcedb8d73e49620a5731196bf7e23bb53ccc

benthecarman commented:
reACK 2e55bcedb8d73e49620a5731196bf7e23bb53ccc

achow101 commented on src/versionbits.cpp:
```diff
@@ -49,20 +56,17 @@ ThresholdState AbstractThresholdConditionChecker::GetStateFor(const CBlockIndex*
         switch (state) {
             case ThresholdState::DEFINED: {
-                if (pindexPrev->GetMedianTimePast() >= nTimeTimeout) {
-                    stateNext = ThresholdState::FAILED;
```
I think that's being checked by the test cases.

sipa replied:
The fuzz test has all transitions of the state machine implemented (in a "come from" form rather than a "go to" one). I don't think it's possible to make incorrect transitions that it won't detect.

achow101 commented on src/test/versionbits_tests.cpp:
```diff
@@ -263,105 +241,72 @@ BOOST_AUTO_TEST_CASE(versionbits_params)
-    const Consensus::Params &mainnetParams = chainParams->GetConsensus();
+    const auto period = CreateChainParams(*m_node.args, CBaseChainParams::REGTEST)->GetConsensus().nMinerConfirmationWindow;
+    gArgs.ForceSetArg("-vbparams", strprintf("testdummy:@%s:@%s", period, period * 3));
+    const auto chainParams = CreateChainParams(*m_node.args, CBaseChainParams::REGTEST);
```
If I need to retouch

achow101 commented on src/chainparams.cpp:
```diff
@@ -474,14 +474,51 @@ class CRegTestParams : public CChainParams {
+    }
+
+    // Actual params must be on retarget block
+    if (startheight < 0) {
```
If I need to retouch

sipa reviewed (commented):
Code review ACK 2e55bcedb8d73e49620a5731196bf7e23bb53ccc

benthecarman commented on src/rpc/blockchain.cpp:
```diff
@@ -1243,8 +1242,8 @@ static void BIP9SoftForkDescPushBack(UniValue& softforks, const std::string &nam
     }
-    bip9.pushKV("start_time", consensusParams.vDeployments[id].nStartTime);
-    bip9.pushKV("timeout", consensusParams.vDeployments[id].nTimeout);
+    bip9.pushKV("startheight", consensusParams.vDeployments[id].startheight);
```
`bip8` or `vbits` since this is no longer `bip9`

sipa replied:
Oh, sure, I thought you were talking about "startheight". This commit only touches observable names, not internal ones.

achow101 replied:
The mass rename is planned for a followup so that this PR is easier to backport.

ajtowns commented:
I think I'm an approach NACK, or at least approach -1, on using heights for activation at this point: they're [incompatible with activating on testnet](https://github.com/bitcoin/bips/pull/1081#pullrequestreview-621934640) due to hash rate inconsistency, and they're incompatible with activation on signet due to custom signets being able to have completely random heights, since anyone can start a new one at any time. They're not especially convenient for activating on mainnet either: we're choosing timeframes and then converting that to expected block heights, but that can have significant errors -- if we want to give people "3 months" to review an update before deploying it, and choose block heights to reflect that, we risk both increased hashpower reducing that time substantially, or reduced hashpower introducing unwanted delays.

The main arguments I've seen for the height based approach is to avoid the signalling phase being skipped entirely, and to prevent a mandatory signalling phase (UASF/bip148/bip91/etc) from being skipped entirely, even in the event of a substantial loss of hashrate. But those goals can be achieved with the MTP approach as well -- see the top commit on https://github.com/ajtowns/bitcoin/commits/202103-bip9-uasf .

There's the additional benefit that it's simpler to explain "signalling begins at height X and ends at height Y" than "signalling begins at the first retarget period after time X, and ends in the last retarget period before time Y", but that seems like a much lower priority than being able to test activations on testnet and signet prior to activation on mainnet...

michaelfolkson commented:
For the sake of respecting other reviewers' time, Taproot is already active on the default Signet for testing and experimentation with Taproot transactions. Activating Taproot on testnet seems like a very low priority, certainly a lower priority than ensuring the optimal code is merged for mainnet activation. 

If you want a summary of the arguments for a consistent use of block height versus using a mix of block height and MTP there is this [SE question](https://bitcoin.stackexchange.com/questions/103854/should-block-height-or-mtp-or-a-mixture-of-both-be-used-in-a-soft-fork-activatio) with two answers, one from harding and one from me summarizing the arguments expressed in the two Speedy Trial PRs. 

This appears to be the major differentiator between the two PRs. If you would prefer a consistent use of block height I would recommend reviewing this PR. If you would prefer a mix of block height and MTP I would recommend reviewing AJ's alternative Speedy Trial PR.

As I've communicated on AJ's PR, spreading review over two competing PRs is doing none of us any favors at this point.

michaelfolkson commented:
Also [this](https://github.com/bitcoin/bitcoin/pull/21377#issuecomment-802336038) shows preferences for a consistent use of block height.

> I don't have a strong view but it appears that @achow101 in #21392 and just by reading the comments in this PR @luke-jr @benthecarman @JeremyRubin @Sjors @roconnor-blockstream @harding all have a preference for a fully height based approach over any use of MTP.

If that isn't the case please correct me. And if anyone else has a view on entirely block height versus some limited use of MTP can you post it on IRC (##taproot-activation)?

The fuzzing PR looks very cool. On that PR @MarcoFalke also expresses an expectation that activation will be based on block heights.

edit: Some additional comments from today's Core dev meeting

"it comes down to mtp vs block height" @achow101

"the main difference between these two PRs are using block height vs MTP " @amitiuttarwar

"with mtp, we run the risk of losing 2 signaling periods. with already few signaling periods, this has the possibility of failing to activate due to bad luck" @achow101

"the property that achow101's PR improves is a fixed (number) of signals" @JeremyRubin

In addition @JeremyRubin stated [here](https://github.com/bitcoin/bitcoin/pull/21377#issuecomment-802396191) "I have a preference for fully height based design, correct."

This is becoming a farce, pure and simple.

benthecarman commented:
> I think I'm an approach NACK, or at least approach -1, on using heights for activation at this point: they're incompatible with activating on testnet due to hash rate inconsistency, and they're incompatible with activation on signet due to custom signets being able to have completely random heights, since anyone can start a new one at any time

I don't think test networks should be relevant to mainnet's activation mechanisms / consensus rules

jnewbery commented on src/versionbits.cpp:
```diff
@@ -49,20 +57,17 @@ ThresholdState AbstractThresholdConditionChecker::GetStateFor(const CBlockIndex*
         vToCompute.pop_back();
 
+        // We track state by previous-block, so the height we should be comparing is +1
+        const int64_t height = pindexPrev->nHeight + 1;
```
Why are you widening height from an `int` (32 bit) to a `int64_t`, and then comparing with `int`s below?

achow101 replied:
Will fix I need to retouch.

jnewbery commented on test/functional/feature_bip8.py:
```diff
@@ -0,0 +1,124 @@
+# Copyright (c) 2021 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+# Test BIP 8 softforks
```
Use docstrings for test description to separate from copyright notice

achow101 replied:
If I need to retouch

jnewbery commented on test/functional/feature_bip8.py:
```diff
@@ -0,0 +1,124 @@
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+# Test BIP 8 softforks
+from collections import namedtuple
+from test_framework.test_framework import BitcoinTestFramework
```
Separate std library imports from local imports

achow101 replied:
If I need to retouch

jnewbery commented on test/functional/feature_bip8.py:
```diff
@@ -0,0 +1,124 @@
+        self.setup_clean_chain = True
+        self.extra_args = [
+            ['-vbparams=testdummy:@-2:@-2'], # Node 0 has TestDummy inactive
+            ['-vbparams=testdummy:@144:@{}'.format(144 * 3)], # Node 1 has regular activation window
```
f-strings are generally preferred in new tests:

```suggestion
            [f'-vbparams=testdummy:@144:@{144 * 3}'], # Node 1 has regular activation window
```

achow101 replied:
These use this format so that it is easier to backport. 0.21 is using a version python that doesn't support f-strings.

jnewbery commented on test/functional/feature_bip8.py:
```diff
@@ -0,0 +1,124 @@
+    def run_test(self):
+        self.log.info("Checking -vbparams")
+        self.stop_node(3)
+        self.nodes[3].assert_start_raises_init_error(extra_args=["-vbparams=testdummy:@-2:@1"], expected_msg="Error: When one of startheight or timeoutheight is -2, both must be -2")
```
There's a lot of repetition here. You could factor this out into a function.

achow101 replied:
It's about as repetitive if this were a function IMO.

jnewbery commented on test/functional/feature_bip8.py:
```diff
@@ -0,0 +1,124 @@
+            self.restart_node(restart.node, restart.extra_args)
+            info = self.nodes[restart.node].getblockchaininfo()
+            assert_equal(info["blocks"], height)
+            if restart.status is None:
+                assert "testdummy" not in info["softforks"]
```
unused

achow101 replied:
If I need to retouch

jnewbery commented on test/functional/feature_bip8.py:
```diff
@@ -0,0 +1,124 @@
+            assert_equal(info["blocks"], height)
+            if status[restart.node] is None:
+                assert "testdummy" not in info["softforks"]
+            else:
+                assert_equal(info["softforks"]["testdummy"]["bip8"]["status"], status[restart.node])
```
unused

achow101 replied:
If I need to retouch

jnewbery commented on src/validation.cpp:
```diff
@@ -1837,10 +1837,11 @@ class WarningBitsConditionChecker : public AbstractThresholdConditionChecker
+    int TimeoutHeight(const Consensus::Params& params) const override { return std::numeric_limits<int>::max(); }
     int Period(const Consensus::Params& params) const override { return params.nMinerConfirmationWindow; }
-    int Threshold(const Consensus::Params& params) const override { return params.nRuleChangeActivationThreshold; }
+    int Threshold(const Consensus::Params& params) const override { return params.m_vbits_min_threshold; }
```
The warning message emitted when this threshold is reached is "Warning: unknown new rules activated", whereas now the threshold being reached really means "Warning: unknown new rules may be activated soon" (since 75% doesn't actually indicate that any vbits deployment will activate).

I think it might make sense to remove the `m_vbits_min_threshold` parameter from chain params and set the threshold to 75% of nMinerConfirmationWindow here.

achow101 replied:
Perhaps for a followup, or if I retouch.

sipa commented:
I believe this is the intended state machine:

![states](https://user-images.githubusercontent.com/548488/112922812-fe08a780-90c1-11eb-99e2-291522811faf.png)

To generate:

```sh
$ dot -Tpng >states.png
digraph versionbits {
    defined [shape=box,label="DEFINED"];
    started [shape=box,label="STARTED"];
    failed [shape=box,label="FAILED"];
    locked [shape=box,label="LOCKED_IN"];
    active [shape=box,label="ACTIVE"];

    defined -> defined [label="height < begin"];
    defined -> started [label="height >= begin"];
    started -> started [label="sig < thresh\nheight < end"];
    started -> failed [label="sig < thresh\nheight >= end"];
    started -> locked [label="sig >= thresh\n"];
    locked -> locked [label="height < min_active"];
    locked -> active [label="height >= min_active"];
    active -> active [label="always"];
    failed -> failed [label="always"];
}
```

Rspigler commented:
Should `STARTED` -> `LOCKED_IN` be defined as "sig >=  thresh &  height  <= end" ?

achow101 commented:
Closing this for now as #21377 (with some changes) is agreeable to most people.
